builder/docker/platform.go

```diff
@@ -28,52 +26,54 @@ import (
 var _ builder.Builder = &dockerBuilder{}
 
 func (b *dockerBuilder) PlatformAdd(opts appTypes.PlatformOptions) error {
-	return b.buildPlatform(opts.Name, opts.Args, opts.Output, opts.Input)
+	return b.buildPlatform(opts.Name, opts.Args, opts.Output, opts.Data)
 }
 
 func (b *dockerBuilder) PlatformUpdate(opts appTypes.PlatformOptions) error {
-	return b.buildPlatform(opts.Name, opts.Args, opts.Output, opts.Input)
+	return b.buildPlatform(opts.Name, opts.Args, opts.Output, opts.Data)
 }
 
-func (b *dockerBuilder) buildPlatform(name string, args map[string]string, w io.Writer, r io.Reader) error {
-	var inputStream io.Reader
-	var dockerfileURL string
-	if r != nil {
-		data, err := ioutil.ReadAll(r)
-		if err != nil {
-			return err
-		}
-		var buf bytes.Buffer
-		writer := tar.NewWriter(&buf)
-		writer.WriteHeader(&tar.Header{
-			Name: "Dockerfile",
-			Mode: 0644,
-			Size: int64(len(data)),
-		})
-		writer.Write(data)
-		writer.Close()
-		inputStream = &buf
-	} else {
-		dockerfileURL = args["dockerfile"]
-		if dockerfileURL == "" {
-			return errors.New("Dockerfile is required")
-		}
-		if _, err := url.ParseRequestURI(dockerfileURL); err != nil {
-			return errors.New("Dockerfile parameter must be a URL")
-		}
+func (b *dockerBuilder) PlatformRemove(name string) error {
+	client, err := getDockerClient()
+	if err != nil {
+		return err
 	}
-	imageName := image.PlatformImageName(name)
+	img, err := client.InspectImage(image.PlatformImageName(name))
+	if err != nil {
+		return err
+	}
+	err = client.RemoveImage(img.ID)
+	if err != nil && err == docker.ErrNoSuchImage {
+		log.Errorf("error removing image %s from Docker: no such image", name)
+		return nil
+	}
+	return err
+}
+
+func (b *dockerBuilder) buildPlatform(name string, args map[string]string, w io.Writer, data []byte) error {
 	client, err := getDockerClient()
 	if err != nil {
 		return err
 	}
+	var inputStream io.Reader
+	var buf bytes.Buffer
+	writer := tar.NewWriter(&buf)
+	writer.WriteHeader(&tar.Header{
```

Comment:
I think we could extract the tar creation logic to a common place shared by the builders